File: infrastructure/web/src/main/java/org/acme/infra/web/rest/CustomerResource.java
package org.acme.infra.web.rest;

import org.acme.app.query.GetCustomers;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.util.List;
import java.util.stream.Collectors;

@Path("/customer")
@Produces(MediaType.APPLICATION_JSON)
public class CustomerResource {

    private GetCustomers getCustomers;

    public CustomerResource(GetCustomers getCustomers) {
        this.getCustomers = getCustomers;
    }

    @GET
    public List<JsonCustomer> list() {
        return getCustomers.getCustomers().stream()
                .map(response -> new JsonCustomer(response.getName()))
                .collect(Collectors.toList());
    }

    public static class JsonCustomer {
        private String name;

        public JsonCustomer(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }
    }
}

File: infrastructure/web/src/main/kotlin/org/acme/infra/web/rest/CustomerResource.kt
package org.acme.infra.web.rest

import jakarta.ws.rs.*
import jakarta.ws.rs.core.Context
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.UriInfo
import org.acme.app.command.CreateCustomerCommand
import org.acme.app.query.GetCustomerQuery
import org.acme.app.query.GetCustomersQuery
import org.acme.infra.web.model.JsonCustomer


@Path("/customer")
@Produces(MediaType.APPLICATION_JSON)
class CustomerResource(
    private val getCustomers: GetCustomersQuery,
    private val getCustomer: GetCustomerQuery,
    private val createCustomerCommand: CreateCustomerCommand,
)
{
    @GET
    fun list(): List<JsonCustomer> {
        return getCustomers.customers {
            it.map { c -> JsonCustomer(c.id, c.name) }
        }

    }

    @GET
    @Path("{id}")
    fun get(@PathParam("id") id: String): JsonCustomer {
        return getCustomer.customer(GetCustomerQuery.Request(id)) {
            JsonCustomer(it.id, it.name)
        }
    }

    @POST
    fun create(payload: CreateCustomerPayload, @Context uriInfo: UriInfo): Response {
        val id = createCustomerCommand.create(CreateCustomerCommand.Request(payload.name)).id
        val uriBuilder = uriInfo.absolutePathBuilder
        uriBuilder.path(id)
        return Response.created(uriBuilder.build()).build()

    }

    data class CreateCustomerPayload(val name: String)
}
